File: frontend/src/app/shared/api/error-envelope.ts
/**
 * Uniform error envelope — mirrors the backend `ErrorEnvelope` record
 * (api-standards.yaml): timestamp / status / error / message / path.
 * Responses are unwrapped; only failures carry this shape.
 */
export interface ErrorEnvelope {
  readonly timestamp: string;
  readonly status: number;
  readonly error: string;
  readonly message: string;
  readonly path: string;
}

/**
 * Typed API error thrown by {@link ApiClient} when the backend returns a
 * non-2xx response. Carries the parsed envelope plus the originating HTTP status.
 */
export class ApiError extends Error {
  constructor(
    readonly envelope: ErrorEnvelope,
    readonly httpStatus: number
  ) {
    super(envelope.message || envelope.error);
    this.name = 'ApiError';
  }
}

File: backend/src/main/java/com/staffengagement/shared/error/ErrorEnvelope.java
package com.staffengagement.shared.error;

import java.time.Instant;

/**
 * Uniform error envelope ({@code api-standards.yaml}). All error responses use this
 * shape: ISO-8601 {@code timestamp}, integer {@code status}, HTTP error name
 * {@code error}, detailed {@code message}, and request {@code path}.
 */
public record ErrorEnvelope(
        Instant timestamp,
        int status,
        String error,
        String message,
        String path) {
}